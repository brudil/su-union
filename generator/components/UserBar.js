import React from 'react';
import MSLTag from './MSLTag';

const loggedInList = (
  <ul className="UserBar__list">
    <li className="UserBar__item UserBar__item-admin UserBar__item--empty UserBar__admin-menu">
      <span>Admin</span>
      <div className="UserBar__item-dropdown" dangerouslySetInnerHTML={{ __html: MSLTag({ name: 'Admin' }) }} />
    </li>
    <li className="UserBar__item UserBar__item-admin UserBar__item--empty UserBar__admin-control-panel">
      <span>Page</span>
      <div className="UserBar__item-dropdown" dangerouslySetInnerHTML={{ __html: MSLTag({ name: 'ControlPanel' }) }} />
    </li>

    <li className="UserBar__item UserBar__item--welcome" />
    <li className="UserBar__item UserBar__item--action">
      <a href="/login" data-action="login">Log in</a>
    </li>

    <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
      <a href="/basket">Basket</a>
    </li>

  </ul>
);

const anonymousList = (
  <ul className="UserBar__list">
    <li className="UserBar__item">Hi there!</li>
    <li className="UserBar__item UserBar__item--action">
      <a href="/login" data-action="login">Log in</a>
    </li>
  </ul>
);


const UserBar = ({ loggedIn }) => (
  <div className="UserBar">
    <div className="Container Container--sv-bleed">
      {loggedIn ? loggedInList : anonymousList}
    </div>
  </div>
);

UserBar.propTypes = {
  loggedIn: React.PropTypes.bool.isRequired,
};

export default UserBar;
