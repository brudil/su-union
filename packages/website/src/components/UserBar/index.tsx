import React, { useState, useCallback, useEffect } from 'react';
import cx from 'classnames';
import client, { ClientAuth } from '@ussu/common/src/libs/user';
import { WebsiteRootState } from '../../types/website';
import { useMappedState } from 'redux-react-hook';

enum DropdownState {
  Page,
  Admin,
}

interface UserBarState {
  isLoaded: boolean;
  dropdownOpen: DropdownState | null;
  auth: ClientAuth | null;
  greetingIndex: 0;
}

export const UserBar: React.FC = ({}) => {
  const [state, setState] = useState<UserBarState>({
    isLoaded: false,
    dropdownOpen: null,
    auth: null,
    greetingIndex: 0,
  });

  const handleToggleDropdown = useCallback(
    (dropdown: DropdownState) => () => {
      setState((state) => ({
        ...state,
        dropdownOpen: dropdown === state.dropdownOpen ? null : dropdown,
      }));
    },
    [],
  );

  const mapState = useCallback(
    (state: WebsiteRootState) => ({
      page: state.page,
      user: state.user,
    }),
    [],
  );

  const {
    page: { menu },
    user: { isLoaded, isLoggedIn, profile, actionBound },
  } = useMappedState(mapState);

  useEffect(() => {
    setTimeout(() => {
      const { auth } = client || { auth: null };
      setState((state) => ({
        ...state,
        auth,
        isLoaded: true,
      }));
    }, 1);

    // trying to avoid using some sort of state container for now. hacky
    (window as any).emitter.on('changePageOptions', (_options: any) => {});
  }, []);

  if (!isLoaded) {
    return (
      <ul className="UserBar__list" style={{ visibility: 'hidden' }}>
        <li className="UserBar__item UserBar__item">Loading</li>
      </ul>
    );
  }

  const { dropdownOpen } = state;

  if (isLoggedIn && profile) {
    return (
      <ul className="UserBar__list">
        <li className="UserBar__item UserBar__item--welcome">
          Hi {profile.firstName}!
        </li>
        <li className="UserBar__item UserBar__item--action">
          <button onClick={actionBound || undefined} type="button">
            Log out
          </button>
        </li>

        {menu.admin.areas.length > 0 || menu.admin.orgs.length > 0 ? (
          <li
            className={cx(
              'UserBar__item UserBar__item-admin UserBar__admin-menu',
              {
                'UserBar__item--open': true,
              },
            )}
          >
            <button
              onClick={handleToggleDropdown(DropdownState.Admin)}
              type="button"
            >
              Admin
            </button>
            {dropdownOpen === DropdownState.Admin ? (
              <div className={cx('UserBar__item-dropdown')}>
                <ul className="UserBar__dropdown-list">
                  {menu.admin.areas.map((item) => (
                    <li key={item.name}>
                      <a href={item.link}>{item.name}</a>
                    </li>
                  ))}
                  <hr />
                  {menu.admin.orgs.map((item) => (
                    <li key={item.name}>
                      <a href={item.link}>{item.name}</a>
                    </li>
                  ))}
                </ul>
              </div>
            ) : null}
          </li>
        ) : null}
        {menu.page.actions.length > 0 ? (
          <li
            className={cx(
              'UserBar__item UserBar__item-admin UserBar__admin-menu',
              {
                'UserBar__item--open': true,
              },
            )}
          >
            <button
              onClick={handleToggleDropdown(DropdownState.Page)}
              type="button"
            >
              Page
            </button>
            {dropdownOpen === DropdownState.Page ? (
              <div
                className={cx('UserBar__item-dropdown', {
                  'UserBar__item--open': dropdownOpen === DropdownState.Page,
                })}
              >
                <ul className="UserBar__dropdown-list">
                  {menu.page.actions.map((item) => (
                    <li key={item.name}>
                      <a href={item.link}>{item.name}</a>
                    </li>
                  ))}
                </ul>
              </div>
            ) : null}
          </li>
        ) : null}

        <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
          <a href="/shop/basket">Basket</a>
        </li>
      </ul>
    );
  }

  return (
    <ul className="UserBar__list">
      <li className="UserBar__item">Hi there!</li>
      <li className="UserBar__item UserBar__item--action">
        <a href="/login">Log in</a>
      </li>
      <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
        <a href="/shop/basket">Basket</a>
      </li>
    </ul>
  );
};
