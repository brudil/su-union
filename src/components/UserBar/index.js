import React from 'react';
import cx from 'classnames';
import ClickOutside from 'react-click-outside';
import currentUser from '~libs/user';
import Hydroleaf from '~components/HydroLeaf';

const PAGE_DROPDOWN = 'page';
const ADMIN_DROPDOWN = 'admin';

class UserBar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoaded: false,
      dropdownOpen: null,
    };

    this.handleOpenDropdown = dropdown =>
      this.setState({ dropdownOpen: dropdown });
    this.handleToggleDropdown = dropdown =>
      this.setState({
        dropdownOpen: dropdown === this.state.dropdownOpen ? null : dropdown,
      });
    this.handleOpenPageDropdown = this.handleOpenDropdown.bind(
      this,
      PAGE_DROPDOWN
    );
    this.handleOpenAdminDropdown = this.handleOpenDropdown.bind(
      this,
      ADMIN_DROPDOWN
    );
    this.handleTogglePageDropdown = this.handleToggleDropdown.bind(
      this,
      PAGE_DROPDOWN
    );
    this.handleToggleAdminDropdown = this.handleToggleDropdown.bind(
      this,
      ADMIN_DROPDOWN
    );
    this.handleCloseDropdown = () => this.setState({ dropdownOpen: null });
  }

  componentDidMount() {
    setTimeout(() => {
      const { auth } = currentUser;
      this.setState({
        isLoaded: true,
        ...auth,
      });
    }, 1);
  }

  renderLoaded() {
    const {
      isLoggedIn,
      details,
      admin,
      page,
      dropdownOpen,
      actionBound,
    } = this.state;

    if (isLoggedIn) {
      return (
        <ul className="UserBar__list">
          <li className="UserBar__item UserBar__item--welcome">
            Hi {details.firstName}!
          </li>
          <li className="UserBar__item UserBar__item--action">
            <button onClick={actionBound} type="button">
              Log out
            </button>
          </li>

          {admin !== null ? (
            <li
              className={cx(
                'UserBar__item UserBar__item-admin UserBar__admin-menu',
                {
                  'UserBar__item--open': dropdownOpen === ADMIN_DROPDOWN,
                }
              )}
            >
              <button onClick={this.handleToggleAdminDropdown} type="button">
                Admin
              </button>
              <ClickOutside onClickOutside={this.handleCloseDropdown}>
                <div className={cx('UserBar__item-dropdown')}>
                  <ul className="UserBar__dropdown-list">
                    {admin.admin.map(item => (
                      <li key={item.name}>
                        <a href={item.link}>{item.name}</a>
                      </li>
                    ))}
                    {admin.orgs.map(item => (
                      <li key={item.name}>
                        <a href={item.link}>{item.name}</a>
                      </li>
                    ))}
                  </ul>
                </div>
              </ClickOutside>
            </li>
          ) : null}
          {page !== null ? (
            <li
              className={cx(
                'UserBar__item UserBar__item-admin UserBar__admin-menu',
                {
                  'UserBar__item--open': dropdownOpen === PAGE_DROPDOWN,
                }
              )}
            >
              <button onClick={this.handleTogglePageDropdown} type="button">
                Page
              </button>
              <ClickOutside onClickOutside={this.handleCloseDropdown}>
                <div
                  className={cx('UserBar__item-dropdown', {
                    'UserBar__item--open': dropdownOpen === PAGE_DROPDOWN,
                  })}
                >
                  <ul className="UserBar__dropdown-list">
                    {page.items.map(item => (
                      <li key={item.name}>
                        <a href={item.link}>{item.name}</a>
                      </li>
                    ))}
                  </ul>
                </div>
              </ClickOutside>
            </li>
          ) : null}

          <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
            <a href="/shop/basket">Basket</a>
          </li>
        </ul>
      );
    }
    return (
      <ul className="UserBar__list">
        <li className="UserBar__item">Hi there!</li>
        <li className="UserBar__item UserBar__item--action">
          <a href="/login" data-action="login">
            Log in
          </a>
        </li>
        <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
          <a href="/shop/basket">Basket</a>
        </li>
      </ul>
    );
  }

  render() {
    const { isLoaded } = this.state;
    if (isLoaded) {
      return this.renderLoaded();
    }

    return (
      <ul className="UserBar__list" style={{ visibility: 'hidden' }}>
        <li className="UserBar__item UserBar__item">Loading</li>
      </ul>
    );
  }
}

export default UserBar;

function DesktopContainer() {
  return (
    <div className="Container UserBar__container">
      <button
        className="UserBar__a11y"
        type="button"
        onClick={() => window.loadRecite()}
      >
        Accessibility
      </button>
      <UserBar />
    </div>
  );
}

export const DesktopUserBar = Hydroleaf({
  className: 'UserBar UserBar--desktop',
  name: 'UserBar',
})(DesktopContainer);
