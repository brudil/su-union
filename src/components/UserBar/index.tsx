import React from 'react';
import cx from 'classnames';
import ClickOutside from 'react-onclickout';
import client, { ClientAuth } from '~libs/user';
import Hydroleaf from '~components/HydroLeaf';

enum DropdownState {
  Page,
  Admin,
}

interface IProps {}

interface IState {
  isLoaded: boolean;
  dropdownOpen: DropdownState | null;
  auth: ClientAuth | null;
}

class UserBar extends React.Component<IProps, IState> {
  private handleToggleDropdown: (dropdown: DropdownState) => void;
  private handleTogglePageDropdown: () => void;
  private handleToggleAdminDropdown: () => void;
  private handleCloseDropdown: (e: React.MouseEvent<HTMLElement>) => void;

  constructor(props: IProps) {
    super(props);

    this.state = {
      isLoaded: false,
      dropdownOpen: null,
      auth: null,
    };

    this.handleToggleDropdown = (dropdown) =>
      this.setState({
        dropdownOpen: dropdown === this.state.dropdownOpen ? null : dropdown,
      });
    this.handleTogglePageDropdown = this.handleToggleDropdown.bind(
      this,
      DropdownState.Page,
    );
    this.handleToggleAdminDropdown = this.handleToggleDropdown.bind(
      this,
      DropdownState.Admin,
    );
    this.handleCloseDropdown = (
      e: React.MouseEvent<HTMLElement | Document>,
    ) => {
      if (
        e.currentTarget.parentElement &&
        e.currentTarget.parentElement.classList.contains('UserBar__item')
      ) {
        return;
      }

      this.setState({ dropdownOpen: null });
    };
  }

  componentDidMount() {
    setTimeout(() => {
      const { auth } = client || { auth: null };
      this.setState((state) => ({
        ...state,
        auth,
        isLoaded: true,
      }));
    }, 1);

    // trying to avoid using some sort of state container for now. hacky
    (window as any).emitter.on('changePageOptions', (options: any) => {
      this.setState((state) => ({
        auth: ({ ...state.auth, page: options } as ClientAuth)
      }));
    })
  }

  renderLoaded() {
    const { auth, dropdownOpen } = this.state;

    if (!auth) {
      return null;
    }

    const { isLoggedIn, profile, admin, page, actionBound } = auth;

    if (isLoggedIn) {
      return (
        <ul className="UserBar__list">
          <li className="UserBar__item UserBar__item--welcome">
            Hi {profile.firstName}!
          </li>
          <li className="UserBar__item UserBar__item--action">
            <button onClick={actionBound} type="button">
              Log out
            </button>
          </li>

          {admin !== null ? (
            <li
              className={cx(
                'UserBar__item UserBar__item-admin UserBar__admin-menu',
                {
                  'UserBar__item--open': true,
                },
              )}
            >
              <button onClick={this.handleToggleAdminDropdown} type="button">
                Admin
              </button>
              {dropdownOpen === DropdownState.Admin ? (
                <ClickOutside onClickOut={this.handleCloseDropdown}>
                  <div className={cx('UserBar__item-dropdown')}>
                    <ul className="UserBar__dropdown-list">
                      {admin.admin.map((item) => (
                        <li key={item.name}>
                          <a href={item.link}>{item.name}</a>
                        </li>
                      ))}
                      <hr />
                      {admin.orgs.map((item) => (
                        <li key={item.name}>
                          <a href={item.link}>{item.name}</a>
                        </li>
                      ))}
                    </ul>
                  </div>
                </ClickOutside>
              ) : null}
            </li>
          ) : null}
          {page !== null ? (
            <li
              className={cx(
                'UserBar__item UserBar__item-admin UserBar__admin-menu',
                {
                  'UserBar__item--open': true,
                },
              )}
            >
              <button onClick={this.handleTogglePageDropdown} type="button">
                Page
              </button>
              {dropdownOpen === DropdownState.Page ? (
                <ClickOutside onClickOut={this.handleCloseDropdown}>
                  <div
                    className={cx('UserBar__item-dropdown', {
                      'UserBar__item--open':
                        dropdownOpen === DropdownState.Page,
                    })}
                  >
                    <ul className="UserBar__dropdown-list">
                      {page.items.map((item) => (
                        <li key={item.name}>
                          <a href={item.link}>{item.name}</a>
                        </li>
                      ))}
                    </ul>
                  </div>
                </ClickOutside>
              ) : null}
            </li>
          ) : null}

          <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
            <a href="/shop/basket">Basket</a>
          </li>
        </ul>
      );
    }
    return (
      <ul className="UserBar__list">
        <li className="UserBar__item">Hi there!</li>
        <li className="UserBar__item UserBar__item--action">
          <a href="/login" data-action="login">
            Log in
          </a>
        </li>
        <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
          <a href="/shop/basket">Basket</a>
        </li>
      </ul>
    );
  }

  render() {
    const { isLoaded } = this.state;
    if (isLoaded) {
      return this.renderLoaded();
    }

    return (
      <ul className="UserBar__list" style={{ visibility: 'hidden' }}>
        <li className="UserBar__item UserBar__item">Loading</li>
      </ul>
    );
  }
}

export default UserBar;

function DesktopContainer() {
  return (
    <div className="Container UserBar__container">
      {/* <button */}
      {/* className="UserBar__a11y" */}
      {/* type="button" */}
      {/* onClick={() => window.loadRecite()} */}
      {/* > */}
      {/* Accessibility */}
      {/* </button> */}
      <UserBar />
    </div>
  );
}

export const DesktopUserBar = Hydroleaf({
  className: 'UserBar UserBar--desktop',
  name: 'UserBar',
})(DesktopContainer);
