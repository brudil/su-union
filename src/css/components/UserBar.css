@import '../core/typography.css';

.UserBar {
  padding: 0.25lh 0;
  background: #f7f7f7;
  color: #000000;
  font-weight: bold;
  font-size: 1.1rem;
  font-family: var(--type__primary);

  &__list {
    padding: 0;
    margin: 0;
    list-style: none;
    display: flex;
    justify-content: flex-end;
  }

  &__item {
    padding-left: 0.5lh;
    margin-right: 0;
  }

  & a {
    color: #000000;
    font-weight: bold;
  }
}
