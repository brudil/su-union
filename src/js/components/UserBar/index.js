import React from 'react';
import cx from 'classnames';
import ClickOutside from 'react-click-outside';
import Hydroleaf from '../HydroLeaf';
import currentUser from '../../libs/user';

const PAGE_DROPDOWN = 'page';
const ADMIN_DROPDOWN = 'admin';

class UserBar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoaded: false,
      dropdownOpen: null,
      isPageOpen: false,
    };

    this.handleOpenDropdown = dropdown =>
      this.setState({ dropdownOpen: dropdown });
    this.handleOpenPageDropdown = this.handleOpenDropdown.bind(
      this,
      PAGE_DROPDOWN
    );
    this.handleOpenAdminDropdown = this.handleOpenDropdown.bind(
      this,
      ADMIN_DROPDOWN
    );
    this.handleCloseDropdown = () => this.setState({ dropdownOpen: null });
  }

  componentDidMount() {
    setTimeout(() => {
      const { auth } = currentUser;
      console.log(currentUser);
      this.setState({
        isLoaded: true,
        ...auth,
      });
    }, 1);
  }

  renderLoaded() {
    const { isLoggedIn, details, admin, page, dropdownOpen } = this.state;

    if (isLoggedIn) {
      return (
        <ul className="UserBar__list">
          <li className="UserBar__item UserBar__item--welcome">
            Hi {details.firstName}!
          </li>
          <li className="UserBar__item UserBar__item--action">[Logout]</li>

          {admin !== null
            ? <li
                className={cx(
                  'UserBar__item UserBar__item-admin UserBar__admin-menu',
                  {
                    'UserBar__item--open': dropdownOpen === ADMIN_DROPDOWN,
                  }
                )}
              >
                <button onClick={this.handleOpenAdminDropdown}>Admin</button>
                <ClickOutside onClickOutside={this.handleCloseDropdown}>
                  <div className={cx('UserBar__item-dropdown')}>
                    <ul>
                      {admin.admin.map(item =>
                        <li>
                          <a href={item.link}>
                            {item.name}
                          </a>
                        </li>
                      )}
                    </ul>
                  </div>
                </ClickOutside>
              </li>
            : null}
          {page !== null
            ? <li
                className={cx(
                  'UserBar__item UserBar__item-admin UserBar__admin-menu',
                  {
                    'UserBar__item--open': dropdownOpen === PAGE_DROPDOWN,
                  }
                )}
              >
                {' '}<button onClick={this.handleOpenPageDropdown}>Page</button>
                <ClickOutside onClickOutside={this.handleCloseDropdown}>
                  <div
                    className={cx('UserBar__item-dropdown', {
                      'UserBar__item--open': dropdownOpen === PAGE_DROPDOWN,
                    })}
                  >
                    <ul>
                      {page.items.map(item =>
                        <li>
                          <a href={item.link}>
                            {item.name}
                          </a>
                        </li>
                      )}
                    </ul>
                  </div>
                </ClickOutside>
              </li>
            : null}

          <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
            <a href="/shop/basket">Basket</a>
          </li>
        </ul>
      );
    }
    return (
      <ul className="UserBar__list">
        <li className="UserBar__item">Hi there!</li>
        <li className="UserBar__item UserBar__item--action">
          <a href="/login" data-action="login">
            Log in
          </a>
        </li>
        <li className="UserBar__item UserBar__item--action UserBar__item--action-highlight">
          <a href="/shop/basket">Basket</a>
        </li>
      </ul>
    );
  }

  render() {
    const { isLoaded } = this.state;

    return (
      <div className="Container UserBar__container">
        {isLoaded
          ? this.renderLoaded()
          : <ul className="UserBar__list" style={{ visibility: 'hidden' }}>
              <li className="UserBar__item UserBar__item">Loading</li>
            </ul>}
      </div>
    );
  }
}

export default Hydroleaf({
  className: 'UserBar',
})(UserBar);
