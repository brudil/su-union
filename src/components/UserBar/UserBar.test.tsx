import React from 'react';
import renderer from 'react-test-renderer';

import UserBar from './index';

it('renders with an anchor', () => {
  const tree = renderer.create(<UserBar />).toJSON();
  expect(tree).toMatchSnapshot();
});
