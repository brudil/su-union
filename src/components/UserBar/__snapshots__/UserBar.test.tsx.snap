// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with an anchor 1`] = `
<ul
  className="UserBar__list"
  style={
    Object {
      "visibility": "hidden",
    }
  }
>
  <li
    className="UserBar__item UserBar__item"
  >
    Loading
  </li>
</ul>
`;
